### PR TITLE
 Add polysemy to spar; promote the SAMLUser CRUD interface to an effect

### PR DESCRIPTION
**Pull request by fisx**
Cloned from #1775 (forks are not picked up by our CI); see description there.

## Checklist

 - [x] The **PR Title** explains the impact of the change.
 - [x] The **PR description** provides context as to why the change should occur and what the code contributes to that effect. This could also be a link to a JIRA ticket or a Github issue, if there is one.
 - [x] **changelog.d** contains the following bits of information:
   - [x] A file with the changelog entry in one or more suitable sub-sections. The sub-sections are marked by directories inside `changelog.d`.

### CODE REVIEW

File: services/spar/src/Spar/App.hs [hunk 1]
```diff
@@ -166,22 +195,35 @@ instance SPStoreIdP SparError Spar where
...
 
+instance Member (Final IO) r => Catch.MonadThrow (Sem r) where
+  throwM = embedFinal . Catch.throwM @IO
+
+instance Member (Final IO) r => Catch.MonadCatch (Sem r) where
+  catch m handler = withStrategicToFinal @IO $ do
+    m' <- runS m
+    st <- getInitialStateS
+    handler' <- bindS handler
+    pure $ m' `Catch.catch` \e -> handler' $ e <$ st
```
Comment (mdimjasevic) [inline services/spar/src/Spar/App.hs:211]: I'm not familiar with strategies, so I don't get the combination of `runS`, `getInitialStateS` and `bindS`. @fisx , do you understand this part?

**Review by mdimjasevic — APPROVED** [review]
Overall, this looks great!

File: services/spar/src/Spar/App.hs (hunk 1 above)
Comment (fisx) [reply services/spar/src/Spar/App.hs:211]: I just found this: https://reasonablypolymorphic.com/blog/tactics/

I don't think it's very important, though: this is something very unlikely to get any bugs past the type system *and* the tests, and the instance will go away as we make more progress with refactoring.  Then again I suspect understanding tactical and strategy may help us understand later versions of this as well.  :)

File: services/spar/src/Spar/App.hs [hunk 2]
```diff
@@ -1,5 +1,7 @@
 {-# LANGUAGE GeneralizedNewtypeDeriving #-}
 {-# LANGUAGE RecordWildCards #-}
+{-# OPTIONS_GHC -fno-warn-orphans #-}
+{-# OPTIONS_GHC -fplugin=Polysemy.Plugin #-}
```
Comment (fisx) [inline services/spar/src/Spar/App.hs:4]: We probably want to add this to the cabal file for the entire project?

File: services/spar/src/Spar/App.hs [hunk 3]
```diff
@@ -95,8 +102,30 @@ import Wire.API.User.IdentityProvider
 import Wire.API.User.Saml
 import Wire.API.User.Scim (ValidExternalId (..))
 
-newtype Spar a = Spar {fromSpar :: ReaderT Env (ExceptT SparError IO) a}
-  deriving (Functor, Applicative, Monad, MonadIO, MonadReader Env, MonadError SparError)
+newtype Spar r a = Spar {fromSpar :: Member (Final IO) r => ReaderT Env (ExceptT SparError (Sem r)) a}
+  deriving (Functor)
+
+liftSem :: (Member (Final IO) r => Sem r a) -> Spar r a
+liftSem r = Spar $ lift $ lift r
```
Comment (fisx) [inline services/spar/src/Spar/App.hs:109]:
```suggestion
liftSem sem = Spar $ lift $ lift sem
```

or, even more wire style:

```suggestion
liftSem = Spar . lift . lift
```

(but it's not important)

File: services/spar/src/Spar/App.hs [hunk 4]
```diff
@@ -166,22 +195,35 @@ instance SPStoreIdP SparError Spar where
...
   Spar $ do
     ctx <- asks sparCtxCas
-    runClient ctx action
+    lift (lift $ embedFinal @IO $ runClient ctx action)
       `Catch.catch` (throwSpar . SparCassandraError . cs . show @SomeException)
 
-insertUser :: SAML.UserRef -> UserId -> Spar ()
-insertUser uref uid = wrapMonadClient $ Data.insertSAMLUser uref uid
+wrapMonadClientSem :: Sem r a -> Spar r a
+wrapMonadClientSem action = liftSem $ action
```
Comment (fisx) [inline services/spar/src/Spar/App.hs:223]: Why the alias, and not just use `liftSem`?

File: services/spar/src/Spar/App.hs (hunk 1 above)
Comment (isovector) [reply services/spar/src/Spar/App.hs:211]: These are low-level poslysemy constructs. Inside `withStrategicToFinal`, `runS` spiritually turns `Sem r a` into `IO a`, by running all *future* interpreters on it. Bind is the same, but for kleisli arrows.

The `getInitialStateS` lets us look at all of the accumulated functorial state up to this point in the program, and manipulate it as we please. For example, if there is a `State s` effect still to be interpreteted, we need to keep track of what state it has!

But @fisx is correct in that we'll very clearly see problems in CI if they are here.

**Review by isovector — APPROVED** [review]

File: services/spar/src/Spar/Sem/SAMLUser/Cassandra.hs [hunk 5]
```diff
@@ -0,0 +1,28 @@
...
+    embed . \case
+      Insert ur uid -> Data.insertSAMLUser ur uid
+      Get ur -> Data.getSAMLUser ur
+      GetAnyByIssuer is -> Data.getSAMLAnyUserByIssuer is
+      GetSomeByIssuer is -> Data.getSAMLSomeUsersByIssuer is
+      DeleteByIssuer is -> Data.deleteSAMLUsersByIssuer is
+      Delete uid ur -> Data.deleteSAMLUser uid ur
+
+interpretClientToIO :: Member (Final IO) r => ClientState -> Sem (Embed Client ': r) a -> Sem r a
+interpretClientToIO ctx = interpret $ \case
```
Comment (fisx) [inline services/spar/src/Spar/Sem/SAMLUser/Cassandra.hs:27]: Naming: should both be have `interpret` in their name?  Or neither?  Or should we just export the composite?

**Review by fisx — COMMENTED** [review]
In summary I'm also happy with this (and thanks @mdimjasevic for taking a look!).

**Comment by fisx** [comment]
I'll merge when CI is done.  (It's obvious the failure was not related to this PR, and integration tests passed locally on my machine, but more green's always nice.)

**Comment by fisx** [comment]
be1ec48  associativity is hard, it seems...

File: services/spar/src/Spar/App.hs (hunk 4 above)
Comment (isovector) [reply services/spar/src/Spar/App.hs:223]: Nice catch --- we unintentionally changed the logic here! This should still be catching the exceptions like `wrapMonadClient` does.

**Pull request merged**